app/Command/Codegen.hs

```diff
@@ -0,0 +1,107 @@
+module Command.Codegen (command) where
+
+import Prelude
+
+import           Command.Common (globWarningOnMisses, printWarningsAndErrors)
+
+import           Control.Applicative (many)
+import           Control.Monad (when, unless)
+import           Control.Monad.Supply
+
+import qualified Data.Aeson.Internal as A
+import           Data.Aeson.Parser (eitherDecodeWith, json)
+import qualified Data.ByteString.Lazy as BSL
+import           Data.Either (lefts, rights)
+import qualified Data.Set as S
+import qualified Data.Map as M
+
+import qualified Language.PureScript as P
+import qualified Language.PureScript.CoreFn as CoreFn
+import qualified Language.PureScript.CoreFn.FromJSON as CoreFn
+
+import qualified Options.Applicative as Opts
+
+import           System.Exit (exitFailure)
+import           System.IO (hPutStr, hPutStrLn, stderr)
+
+data CodegenOptions = CodegenOptions
+  { codegenCoreFnInput :: [FilePath]
+  , codegenJSONErrors :: Bool
+  , codegenOutputDir :: FilePath
+  }
+
+codegen :: CodegenOptions -> IO ()
+codegen CodegenOptions{..} = do
+  inputFiles <- globWarningOnMisses (unless codegenJSONErrors . warnFileTypeNotFound) codegenCoreFnInput
+  when (null inputFiles && not codegenJSONErrors) $ do
+    hPutStr stderr $ unlines
+      [ "purs codegen: No input files."
+      , "Usage: For basic information, try the `--help` option."
+      ]
+    exitFailure
+  mods <- traverse parseCoreFn inputFiles
+
+  let
+    filePathMap =
+      M.fromList $ map ((\m -> (CoreFn.moduleName m, Right $ CoreFn.modulePath m)) . snd) $ rights mods
+
+  unless (null (lefts mods)) $ do
+    _ <- traverse (hPutStr stderr . formatParseError) $ lefts mods
+    exitFailure
+
+  foreigns <- P.inferForeignModules filePathMap
+  (makeResult, makeWarnings) <-
+    P.runMake purescriptOptions
+      $ runSupplyT 0
```

Comment:
In the normal compilation path, the codegen supply monad is initialized with the next unused number from previous parts of the compilation. Using 0 here raises the question of whether this reuse is necessary. If so, using 0 here might cause problems. If not (I suspect not), we should probably be consistent so that the produced code isn't different when generated just by `purs compile` versus `purs compile; purs codegen`.

So assuming it's safe to do so, I think we should remove the `SupplyT` from the signatures in `MakeActions` and push that detail into their implementations. But now would be a really good time for someone else to share why that wouldn't be safe!

Reply:
This is a great point. I tried reading through the usages of the supply monad in the codegen code and it looks like it's just for generating fresh variable names - it doesn't seem to me that it'd require we start off from where we left off in say typechecking - but I don't have enough experience to say for sure.

At work we've been using `zephyr` for quite a while, which also starts from 0 for codegen, so I'd be really surprised if it causes errors.

If it is the case that it doesn't matter, then I'll remove the `SupplyT` requirement and start it from zero within `codegenJS`.

Reply:
I spent some time going through the `codegenJS` implementation and I don't think that it is dangerous to always initial that supply with 0. The fact that [zephyr was doing that](https://github.com/coot/zephyr/blob/master/app/Command/Run.hs#L296) for so long also makes me pretty confident based on my personal experience.

I went ahead and made the change you suggested, and all tests are passing.

If anyone knows more than I do and thinks that we should undo the change, I can do that too!